source/CIS/cis-3.rst

```diff
@@ -0,0 +1,242 @@
+pays for the transaction fees while the sponsoree has to sign its intended action in a message to authorize the sponsor to
+execute a specific action on its behalf.
+
+Additionally, the interface provides an entrypoint for querying the smart contract about which
+entrypoints can be invoked with a sponsored transaction.
+The smart contract responds with either: not supported, or supported.
+
+Specification
+=============
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in :rfc:`2119`.
+
+General types and serialization
+-------------------------------
+
+.. _CIS-3-ContractAddress:
+
+``ContractAddress``
+^^^^^^^^^^^^^^^^^^^
+
+An address of a contract instance.
+It consists of an index and a subindex both unsigned 64 bit integers.
+
+It is serialized as: First 8 bytes for the index (``index``) followed by 8 bytes for the subindex (``subindex``) both little endian::
+
+  ContractAddress ::= (index: Byte⁸) (subindex: Byte⁸)
+
+.. _CIS-3-AccountAddress:
+
+``AccountAddress``
+^^^^^^^^^^^^^^^^^^
+
+An address of an account.
+
+It is serialized as 32 bytes::
+
+  AccountAddress ::= (address: Byte³²)
+
+.. _CIS-3-OwnedEntrypointName:
+
+``OwnedEntrypointName``
+^^^^^^^^^^^^^^^^^^^^^^^
+
+A name for a smart contract function entrypoint (owned version).
+
+It is serialized as: First 2 bytes encode the length (``n``) of the entrypoint name, followed by this many bytes for the entrypoint name (``entrypoint``)::
+
+  OwnedEntrypointName ::= (n: Byte²) (entrypoint: Byteⁿ)
+
+.. _CIS-3-Timestamp:
+
+``Timestamp``
+^^^^^^^^^^^^^
+
+A timestamp given in milliseconds since unix epoch.
+It consists of an unsigned 64 bit integer.
+
+It is serialized as 8 bytes in little endian::
+
+  Timestamp ::= (milliseconds: Byte⁸)
+
+.. _CIS-3-Nonce:
+
+``Nonce``
+^^^^^^^^^
+
+An unsigned 64 bit integer number that increases sequentially to protect against replay attacks.
+
+It is serialized as 8 bytes in little endian::
+
+  Nonce ::= (nonce: Byte⁸)
+
+.. _CIS-3-SignatureEd25519:
+
+``SignatureEd25519``
+^^^^^^^^^^^^^^^^^^^^^^
+
+Signature for an Ed25519 message.
+
+It is serialized as 64 bytes::
+
+  SignatureEd25519 ::= (signature: Byte⁶⁴)
+
+.. _CIS-3-TwoLevelSignatureMap:
+
+``TwoLevelSignatureMap``
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Two-level map based on B-Trees. A BTreeMap stores key-value pairs, where the keys are unsigned 8 bit integers in both maps.
+The value in the other map is the inner map and the value in the inner map is of type ``SignatureEd25519``.
+
+It is serialized as: For each map, the first 4 bytes encode the size (``n`` for the other map and ``m`` for the inner map)
+of the map, followed by this many key-value pairs::
+
+  TwoLevelSignatureMap ::= (n: Byte⁴) (key: Byte, value: (m: Byte⁴) (key: Byte, value: SignatureEd25519)ᵐ)ⁿ
```

Comment:
Since the keys are only 1 byte each, we don't need 4 bytes to encode the length of each of the maps.
```suggestion
  TwoLevelSignatureMap ::= (n: Byte) (key: Byte, value: (m: Byte) (key: Byte, value: SignatureEd25519)ᵐ)ⁿ
```

Reply:
``n`` and ``m`` are the sizes of the map and I got the info from this line that `u32` is used for the size of the BTreeMap size:
https://github.com/Concordium/concordium-contracts-common/blob/main/concordium-contracts-common/src/impls.rs#L1017

Reply:
Right, so that is why we sometimes add `#[concordium(size_length = 1)]` like here https://github.com/Concordium/concordium-rust-smart-contracts/blob/697dad29e60ec855124709a7f4de5dd8e78da9fe/concordium-cis2/src/lib.rs#L120
It changes the size used to encode the length

Reply:
We have a brief description of this here https://docs.rs/concordium-std/latest/concordium_std/derive.Serial.html

Reply:
Right, that makes a lot of sense now.